src/Text/Parsing/Parser/Combinators.purs

```diff
@@ -168,51 +184,129 @@ sepEndBy1Rec p sep = do
 -- |
 -- | For example:
 -- |
 -- | ```purescript
--- | chainr digit (string "+" *> add) 0
+-- | chainr digit (string "+" $> add) 0
 -- | ```
 chainr :: forall m s a. Monad m => ParserT s m a -> ParserT s m (a -> a -> a) -> a -> ParserT s m a
 chainr p f a = chainr1 p f <|> pure a
 
+-- | Stack-safe version of `chainr` at the expense of `MonadRec` constraint.
+chainrRec :: forall m s a. MonadRec m => ParserT s m a -> ParserT s m (a -> a -> a) -> a -> ParserT s m a
+chainrRec p f a = chainr1Rec p f <|> pure a
+
 -- | Parse phrases delimited by a left-associative operator.
+-- |
+-- | For example:
+-- |
+-- | ```purescript
+-- | chainr digit (string "+" $> add) 0
+-- | ```
 chainl :: forall m s a. Monad m => ParserT s m a -> ParserT s m (a -> a -> a) -> a -> ParserT s m a
 chainl p f a = chainl1 p f <|> pure a
 
+-- | Stack-safe version of `chainl` at the expense of `MonadRec` constraint.
+chainlRec :: forall m s a. MonadRec m => ParserT s m a -> ParserT s m (a -> a -> a) -> a -> ParserT s m a
+chainlRec p f a = chainl1Rec p f <|> pure a
+
 -- | Parse phrases delimited by a left-associative operator, requiring at least one match.
 chainl1 :: forall m s a. Monad m => ParserT s m a -> ParserT s m (a -> a -> a) -> ParserT s m a
 chainl1 p f = do
   a <- p
-  chainl1' p f a
-
-chainl1' :: forall m s a. Monad m => ParserT s m a -> ParserT s m (a -> a -> a) -> a -> ParserT s m a
-chainl1' p f a =
-  ( do
-      f' <- f
-      a' <- p
-      chainl1' p f (f' a a')
-  ) <|> pure a
+  chainl1' a
+  where
+  chainl1' a =
+    ( do
+        f' <- f
+        a' <- p
+        chainl1' (f' a a')
+    ) <|> pure a
+
+-- | Stack-safe version of `chainl1` at the expense of `MonadRec` constraint.
+chainl1Rec :: forall m s a. MonadRec m => ParserT s m a -> ParserT s m (a -> a -> a) -> ParserT s m a
+chainl1Rec p f = do
+  a <- p
+  tailRecM go a
+  where
+  go :: a -> ParserT s m (Step a a)
+  go a =
+    ( do
+        op <- f
+        a' <- p
+        pure $ Loop $ op a a'
+    )
+      <|> pure (Done a)
 
 -- | Parse phrases delimited by a right-associative operator, requiring at least one match.
 chainr1 :: forall m s a. Monad m => ParserT s m a -> ParserT s m (a -> a -> a) -> ParserT s m a
 chainr1 p f = do
   a <- p
-  chainr1' p f a
-
-chainr1' :: forall m s a. Monad m => ParserT s m a -> ParserT s m (a -> a -> a) -> a -> ParserT s m a
-chainr1' p f a =
-  ( do
-      f' <- f
-      a' <- chainr1 p f
-      pure $ f' a a'
-  ) <|> pure a
+  chainr1' a
+  where
+  chainr1' a =
+    ( do
+        f' <- f
+        a' <- chainr1 p f
+        pure $ f' a a'
+    ) <|> pure a
+
+-- | Stack-safe version of `chainr1` at the expense of `MonadRec` constraint.
+chainr1Rec :: forall m s a. MonadRec m => ParserT s m a -> ParserT s m (a -> a -> a) -> ParserT s m a
+chainr1Rec p f = do
+  a <- p
+  tailRecM go { last: a, init: Nil }
+  where
+  -- This looks scary at first glance, so I'm leaving a comment in a vain
```

Comment:
I love this explanation, thanks very much for this.

Reply:
👍